pkg/engine/plan/plan.go

```diff
@@ -402,6 +418,79 @@ func (v *Visitor) AllowVisitor(kind astvisitor.VisitorKind, ref int, visitor int
 		return true
 	}
 	for _, config := range v.planners {
+		// In the following query:
```

Comment:
This is a super-detailed comment!  I'm still not sure I entirely understand it.  (I wonder if it would benefit from saying *why* the pets planner needs to iterate as deep as it does.)  But the length of the comment itself is a useful signal that something tricky is going on. :-)